extensions/amp-analytics/0.1/visibility-impl.js

```diff
@@ -159,27 +159,32 @@ export function getElement(ampdoc, selector, el, selectionMethod) {
     return null;
   }
 
+  const elWin = el.ownerDocument.defaultView;
   // Special case for root selector.
   if (selector == ':host' || selector == ':root') {
-    const elWin = ampdoc.win;
     const parentEl = elWin.frameElement && elWin.frameElement.parentElement;
     if (parentEl) {
       return closestBySelector(parentEl, '.-amp-element');
     }
   }
 
+  let foundEl;
   if (selectionMethod == 'closest') {
     // Only tag names are supported currently.
-    const closestEl = closestByTag(el, selector);
-    // Restrict result to be contained by ampdoc.
-    if (closestEl && ampdoc.contains(closestEl)) {
-      return closestEl;
-    }
-    return null;
+    foundEl = closestByTag(el, selector);
   } else if (selectionMethod == 'scope') {
-    return el.parentElement.querySelector(selector);
+    foundEl = el.parentElement.querySelector(selector);
   } else if (selector[0] == '#') {
-    return ampdoc.getElementById(selector.slice(1));
+    foundEl = el.ownerDocument.getElementById(selector.slice(1));
```

Comment:
Yes, this is not entirely correct. The situations you have to consider here are:

1. The `el` is in the `ampdoc` directly. In that case the answer is simply `ampdoc.getElementById()`.
2. If the `el` is in the friendly iframe, then it should be `ampdoc.getElementById()`.

Whether it's option 1 or 2 you can tell by the presence of `parentEl` above. You probably want to refactor it a bit to use throughout.

Eventually we'll need to build out the math for all this logic, but let's first see what it is here.

Finally: this logic is not so much hard as it just has to deal with multiple environments. It's important to present them all in tests. Right now, from what I can tell, the tests are insufficient. Let's add more.